### DISCUSSION TITLE
Viseron exit code 256 - AMD Athlon(tm) 64 X2 Dual Core Processor 5200

### DISCUSSION THREAD

**Discussion by sgoldtho** [roflcoopter/viseron#690] [Q&A] [answered]
I'm running Viseron in Docker on an Unraid installation and the motion detection works fine, but if I enable object detection I get exit code 256...

Is the CPU too old for this?

[2023-12-06 19:40:55] [INFO    ] [viseron.components] - Setup of component webserver took 0.0 seconds
[2023-12-06 19:40:55] [INFO    ] [viseron.components] - Setting up component nvr
[2023-12-06 19:40:55] [INFO    ] [viseron.components] - Setting up component edgetpu
[2023-12-06 19:40:55] [INFO    ] [viseron.components] - Setup of component nvr took 0.0 seconds
[2023-12-06 19:40:55] [INFO    ] [viseron.components] - Setting up component mog2
[2023-12-06 19:40:55] [INFO    ] [viseron.components] - Setting up component ffmpeg
[viseron-finish] Viseron exit code 256
[viseron-finish] Viseron received signal 4
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.

**Comment by sgoldtho** [comment]
Are OpenCL, VA-API or CUDA needed when there is an Coral USB device available?

************************ UID/GID *************************
User uid:    0
User gid:    0
************************** Done **************************
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-gid-video-device: executing... 
[cont-init.d] 20-gid-video-device: exited 0.
[cont-init.d] 30-edgetpu-permission: executing... 
************** Setting EdgeTPU permissions ***************
Coral Vendor IDs:
"1a6e"
"18d1"
Matching USB devices:
Bus 002 Device 003: ID 1a6e:089a  
No EdgeTPU PCI device was found
************************** Done **************************
[cont-init.d] 30-edgetpu-permission: exited 0.
[cont-init.d] 40-set-env-vars: executing... 
****** Checking for hardware acceleration platforms ******
OpenCL cannot be used
VA-API cannot be used
CUDA cannot be used
*********************** Done *****************************

**Reply by roflcoopter** [reply]
No they are not required. If they are available they will be used for hardware accelerated decoding of camera streams, not for object detection

You could collect the crash dump by mounting /var/crash into the container, then we can see why you get the exit 256 error

**Reply by sgoldtho** [reply]
Hmm okay not sure how to do that...

I have added '-v /var/crash:/var/crash' to the Extra parameters in the Unraid container settings, is that correct

**Reply by sgoldtho** [reply]
[docker-run.pdf](https://github.com/roflcoopter/viseron/files/13606896/docker-run.pdf)
[Viseron-log.pdf](https://github.com/roflcoopter/viseron/files/13606904/Viseron-log.pdf)

Not sure if these help, but I've attached the docker run settings and a log file...

**Reply by roflcoopter** [reply]
Yes that is correct.
However, that is how you would troubleshoot this on a Linux server, not sure how to do this on Unraid so you will have to look around a bit to find how to read the crash logs on Unraid

**Reply by sgoldtho** [reply]
Thanks, I cannot find any way to get crash logs out of the container...

I don't see a directory /var/crash in the Viseron container when it is running without object detection, should there be one?

I see that the device 1a6e:089a has been detected is that correct or does it need to be 18d1:9302 Google Inc?

**Reply by roflcoopter** [reply]
I have never used Unraid so unfortunately im not sure how the crash dumps are processed :(

/var/crash only gets populated when a sysdump occurs (at least on Ubuntu) so it should not appear unless there is an error.

The Coral can have either 1a6e:089a or 18d1:9302 so that should be okay

**Comment by bsyomov** [comment]
Try to use dmesg command in host system after error. 
Unraid is just Linux with systemd, so coredumpctl can help too.

**Comment by sgoldtho** [comment]
Thanks

It appears that coreumpctl is not available to me (command not found).

I tried running Viseron with one other the other sections enable (not both at once):

1) If I enable this section in the config file:

edgetpu:
  object_detector:
    cameras:
      camera_1:
        fps: 2
        labels:
          - label: person
            confidence: 0.8
            trigger_recorder: true

  image_classification:
    device: usb

I see this with dmesg

traps: python3[25986] trap invalid opcode ip:1513f0e1a90c sp:1513f1b18bb0 error:0 in _pywrap_tensorflow_interpreter_wrapper.cpython-38-x86_64-linux-gnu.so[1513f0e00000+361000]



2) If I enable this section of the config file:

darknet:
  object_detector:
    cameras:
      camera_1: # Attach detector to the configured camera_1 above
        fps: 5
        scan_on_motion_only: true # Scan for objects even when there is no motion
        log_all_objects: false
        labels:
          - label: person
            confidence: 0.75
            trigger_recorder: true

I see this with dmesg

traps: python3[17699] trap invalid opcode ip:14c790b1d049 sp:14c77e7f7870 error:0 in libopenvino_intel_cpu_plugin.so[14c79016a000+271c000]


Regards, 
Steve 

If you have received this message in error, please reply to this email to notify the sender of its incorrect delivery, then delete both it and your reply. Thank you.

**Comment by bsyomov** [comment]
I had a similar problem with dlib on an AMD fm2 cpu. Rebuilding dlib solved the problem.
In your case, you can probably try to build container directly on the target system, to use with restricted instruction set of your CPU. (clone repository, build, etc.). However, this is not an easy way. Some programing skill needed. =)

**Comment by bsyomov** [comment]
Also, you can try https://viseron.netlify.app/components-explorer/components/codeprojectai as object detector.

**Comment by sgoldtho** [comment]
Thanks,

If I enable this section of the config file:

codeprojectai:
  host: 192.168.139.118
  port: 32168
  object_detector:
    cameras:
      camera_1:
        fps: 1
        log_all_objects: true
        labels:
          - label: person
            confidence: 0.8
            trigger_recorder: true

I see this with dmesg

traps: python3[13462] trap invalid opcode ip:14e3a3477cde sp:14e3c1d83910 error:0 in _dlib_pybind11.cpython-38-x86_64-linux-gnu.so[14e3a3446000+849000]


Looks like I will need to try and figure out how to try to build container directly on my Unraid machine...

**Reply by bsyomov** [reply]
This is exactly the error I wrote about above.
I use this hack to solve it:

My Dockerfile looks like this:
```
FROM roflcoopter/amd64-cuda-viseron:dev
RUN apt-get -y update && apt-get install -y build-essential cmake python3-dev
RUN pip install --force-reinstall dlib
```
docker-compose.yml changes to use this Dockerfile:
```
  viseron:
#    image: roflcoopter/amd64-cuda-viseron:dev
    build: .
```
It rebuilds Dlib inplace.
You may need to use non cuda image.

**Comment by sgoldtho** [comment]
Thanks

I’ll give that a try, all I need to do is convert that into something that Unraid and it’s Docker understands (not sure what the package manager is for Unraid)…

Steve

**Reply by bsyomov** [reply]
How do you run viseron now? `docker-compose up`? `docker run` command? Some web ui?

`RUN apt-get -y update && apt-get install -y build-essential cmake python3-dev`
It runs in viseron container, so it's independed from unraid package manager.

**Comment by sgoldtho** [comment]
Unraid has an Apps Ui where you download the the app you want and then a Docker web UI where you configure and run the app…

I’ve had a look on Unraid and apt-get, cmake pip are all an unknown command.

I’m starting to think I should move Viseron over to a Raspberry Pi with Docker...

 
Thanks and Regards, 
Steve 

If you have received this message in error, please reply to this email to notify the sender of its incorrect delivery, then delete both it and your reply. Thank you.

**Reply by bsyomov** [reply]
> I’ve had a look on Unraid and apt-get, cmake pip are all an unknown command.

This commands not needed on unraid itself. Build runs inside viseron container.

You can use docker command in CLI to build and run container.
place Dockerfile somwere, then run commands:
```
cd /path/to/your/Dockerfile
run docker build . --tag viseron-hacked
run docker run --rm \
  -v {recordings path}:/recordings \
  -v {config path}:/config \
  -v /etc/localtime:/etc/localtime:ro \
  -p 8888:8888 \
  --name viseron \
  viseron-hacked
```

Also, you can use `docker exec` to run commands in allready running viseron container. Note: remove object detection from configuration and restart container, container must be in running state.

Get viseron container name/id:
`docker ps | grep viseron`

Run build in container:
```
docker exec -it name-or-id-of-viseron-container apt-get -y update
docker exec -it name-or-id-of-viseron-container apt-get install -y build-essential cmake python3-dev
docker exec -it name-or-id-of-viseron-container pip install --force-reinstall dlib
docker restart name-or-id-of-viseron-container
```

**Comment by roflcoopter** [comment]
What CPU is your host running on?

@bsyomov thank you for chiming in!

**Comment by sgoldtho** [comment]
AMD Athlon(tm) 64 X2 Dual Core Processor 5200

Steve 

If you have received this message in error, please reply to this email to notify the sender of its incorrect delivery, then delete both it and your reply. Thank you.

**Reply by roflcoopter** [reply, marked as answer]
Ahh yes that is a bit old, i think its crashing since your CPU does not support AVX2 for instance
As @bsyomov said its possible to build the containers yourself, it takes some time bit its not very hard.

You can follow these steps to build everything: `https://viseron.netlify.app/docs/developers/docker`
